ext/OptimizationDIExt.jl

```diff
@@ -0,0 +1,181 @@
+module OptimizationDIExt
+       
+
+import OptimizationBase, OptimizationBase.ArrayInterface
+import OptimizationBase.SciMLBase: OptimizationFunction
+import OptimizationBase.LinearAlgebra: I
+import DifferentiationInterface
+import DifferentiationInterface: prepare_gradient, prepare_hessian, prepare_jacobian, gradient!!, hessian!!, jacobian!!
+using ADTypes
+
+function OptimizationBase.instantiate_function(f, x, adtype::ADTypes.AbstractADType, p = SciMLBase.NullParameters(), num_cons = 0)
+    _f = (θ, args...) -> first(f.f(θ, p, args...))
+
+    if f.grad === nothing
+        extras_grad = prepare_gradient(_f, adtype, x)
+        function grad(res, θ)
+            gradient!!(_f, res, adtype, θ, extras_grad)
+        end
+    else
+        grad = (G, θ, args...) -> f.grad(G, θ, p, args...)
+    end
+
+    hess_sparsity = f.hess_prototype
+    hess_colors = f.hess_colorvec
+    if f.hess === nothing
+        extras_hess = prepare_hessian(_f, DifferentiationInterface.SecondOrder(adtype), x) #placeholder logic, can be made much better
+        function hess(res, θ, args...)
+            hessian!!(_f, res, adtype, θ, extras_hess)
+        end
+    else
+        hess = (H, θ, args...) -> f.hess(H, θ, p, args...)
+    end
+
+    if f.hv === nothing
+        hv = function (H, θ, v, args...)
+            res = zeros(length(θ), length(θ))
+            hess(res, θ, args...)
+            H .= res * v
+        end
```

Comment:
why not use the `hvp` from DI?

Reply:
yup I will use that this is legacy code from my zygote implementation or something